File: python/day2/__main__.py
import numpy as np
from typing import Callable

def read(filename):
    ranges = []
    with open(filename, "r") as f:
        contents = f.read()
        for id_range in contents.split(","):
            sep_index = id_range.index("-")
            ranges.append((int(id_range[:sep_index]), int(id_range[sep_index+1:])))
    return ranges

def invalid1(n):
    # does the number contain an even number of digits?
    digits = int(np.floor(np.log10(n)) + 1)
    if digits%2 == 1:
        return False

    # split in half and compare
    e = 10 ** (digits//2)
    return (n // e) == (n % e)

def invalid2(n):
    # strategy here is to determine if the
    digits = int(np.floor(np.log10(n)) + 1)
    for d in range(1, digits//2+1):
        # skip if digit count to be repeated doesn't divide total digits evenly
        if digits % d != 0:
            continue

        e = 10**d # magnitude to trim by
        cp = n%e  # trim to least d digits for copying

        # compute the candidate by copying the copy value
        candidate = 0
        for nn in range(0, digits, d):
            candidate += cp * 10**nn
        if candidate == n:
            return True
    return False

def invalid1_seek(r):
    # finds invalud nubmers based on range bounds
    digits = lambda n: int(np.log10(n) + 1)
    low_digits, high_digits = digits(r[0]), digits(r[1])
    for d in range(low_digits, high_digits+1):
        if d%2 == 1:
            continue

        # cut the lower d//2 digits off the lower and upper bounds
        e = 10 ** (d//2)
        low, high = r[0]//e, r[1]//e

        # for every number in this range we just append itself to the end and see if it falls in the original range
        for n in range(max(low, 1), high+1):
            if digits(n)*2 != d:
                continue
            candidate = n*e + n
            if r[0] <= candidate <= r[1]:
                yield candidate

def invalid2_seek(r):
    pass

def puzzle(ranges, invalid_fn: Callable[[int], bool]):
    count=0
    for r in ranges:
        for i in range(r[0], r[1]+1):
            if invalid_fn(i):
                count += i
    return count

def puzzle1_seek(ranges):
    count=0
    for r in ranges:
        for n in invalid1_seek(r):
            count += n
    return count

input = read("input/day2/puzzle.txt")
print("puzzle 1:        ", puzzle(input, invalid1))
print("puzzle 2:        ", puzzle(input, invalid2))
# print("puzzle 1 (fast): ", puzzle1_seek(input))

File: go/day2/main.go
package main

import (
	"fmt"
	"iter"
	"math"
	"os"
	"strconv"
	"strings"
)

func main() {
	ranges := read("input/day2/sample.txt")
	fmt.Println("puzzle1          :", puzzle(ranges, invalid1))
	fmt.Println("puzzle1 generator:", puzzle_generate(ranges, generate1))
	fmt.Println()
	fmt.Println("puzzle2          :", puzzle(ranges, invalid2))
	fmt.Println("puzzle2 generator:", puzzle_generate(ranges, generate2))
}

func puzzle(ranges [][2]int, validFn func(int) bool) int {
	sum := 0
	for _, r := range ranges {
		for n := r[0]; n <= r[1]; n++ {
			if validFn(n) {
				sum += n
			}
		}
	}
	return sum
}

func puzzle_generate(ranges [][2]int, generator func(r [2]int) iter.Seq[int]) int {
	sum := 0
	for _, r := range ranges {
		for n := range generator(r) {
			sum += n
		}
	}
	return sum
}

func invalid1(n int) bool {
	digits := _digits(n)
	if digits%2 != 0 {
		return false
	}
	e := int(math.Pow10(digits / 2))
	return n/e == n%e
}

func invalid2(n int) bool {
	digits := _digits(n)
	for d := 1; d <= digits/2; d++ {
		if digits%d != 0 {
			continue
		}
		e := int(math.Pow10(d))
		if n == buildCandidate(n%e, d, digits, e) {
			return true
		}
	}
	return false
}

// generate1 generates invalid candidates for puzzle 1 based on the range
func generate1(r [2]int) iter.Seq[int] {
	lowDigits, highDigits := _digits(r[0]), _digits(r[1])

	return func(yield func(int) bool) {
		for d := lowDigits; d <= highDigits; d++ {
			// ignore digit counts not an even number
			if d%2 != 0 {
				continue
			}

			// compute low and high ranges for candidate halves
			e := int(math.Pow10(d / 2))
			low, high := r[0]/e, r[1]/e

			for n := max(low, 1); n <= min(high, e-1); n++ {
				candidate := n*e + n
				if inRange(r, candidate) {
					if !yield(candidate) {
						return
					}
				}
			}
		}
	}
}

// generate2 generates invalid candidates for puzzle 1 based on the range
func generate2(r [2]int) iter.Seq[int] {
	lowDigits, highDigits := _digits(r[0]), _digits(r[1])
	seen := make(map[int]bool)
	return func(yield func(int) bool) {
		for d := lowDigits; d <= highDigits; d++ {
			for dd := 1; dd <= d/2; dd++ {
				// ignore digit counts that do not divide the target digits
				if d%dd != 0 {
					continue
				}
				e := int(math.Pow10(dd))
				low, high := r[0]/e, r[1]/e
				for n := max(low, 1); n <= high; n++ {
					nn := n
					for nn >= e {
						nn /= e
					}
					candidate := buildCandidate(nn, dd, d, e)
					if !seen[candidate] && inRange(r, candidate) {
						if !yield(candidate) {
							return
						}
						seen[candidate] = true
					}
				}
			}
		}
	}
}

func read(file string) [][2]int {
	content, err := os.ReadFile(file)
	if err != nil {
		panic(err)
	}
	split := strings.Split(string(content), ",")
	result := make([][2]int, 0, len(split))
	for _, r := range split {
		lowHigh := strings.Index(r, "-")
		lowS, highS := r[:lowHigh], r[lowHigh+1:]
		low, _ := strconv.Atoi(lowS)
		high, _ := strconv.Atoi(highS)
		result = append(result, [2]int{low, high})
	}
	return result
}

func _digits(n int) int {
	return int(math.Ceil(math.Log10(float64(n))))
}

func buildCandidate(n, digits, totalDigits, e int) int {
	candidate := 0
	exp := 1
	for d := 0; d < totalDigits/digits; d++ {
		candidate += n * exp
		exp *= e
	}
	return candidate
}

func inRange(r [2]int, candidate int) bool {
	return r[0] <= candidate && candidate <= r[1]
}
